network/wsPeer.go

```diff
@@ -440,6 +476,15 @@ func (wp *wsPeer) readLoop() {
 		msg.processing = wp.processed
 		msg.Received = time.Now().UnixNano()
 		msg.Data = slurper.Bytes()
+		if msg.Tag == protocol.ProposalPayloadTag &&
+			wp.features&vfCompressedProposal != 0 &&
+			bytes.Equal(msg.Data[:4], zstdCompressionMagic[:]) {
```

Comment:
to simplify this you could introduce a new message tag CompressedProposalPayloadTag?

Reply:
Or if the feature is supported, assume it is always compressed?

Reply:
since this is a network protocol feature, you might have a relay on an old proto, and a node on new, and another way around... although you are right and `wp` is remote side, so we can just check features and do not look into data.

Reply:
you could make it a network protocol feature, you could also keep the network protocol version the same and add it as a HTTP header capability that gets advertised at connect time, either way you will know when you are talking to a wsPeer whether they support the new payload encoding. I am assuming a node supporting the new feature will compress uncompressed payloads it receives before relaying them..

Reply:
done

Reply:
I re-evaluated http-header features vs protocol version features. The former makes it almost impossible to manage batch (arrays) sending for peers supporting different set of features - basically need to prepare unique batches for each type of peers (by feature), and the more features, the more different batches.
The solution is either to retire batches/arrays (BroadcastArray) or limit number of features, or use network protocol-level versions as features aggregate.